src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DoNotUseCompilationGetSemanticModelAnalyzer.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis.Analyzers.MetaAnalyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class DoNotUseCompilationGetSemanticModelAnalyzer : DiagnosticAnalyzerCorrectnessAnalyzer
+    {
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(CodeAnalysisDiagnosticsResources.DoNotUseCompilationGetSemanticModelTitle), CodeAnalysisDiagnosticsResources.ResourceManager, typeof(CodeAnalysisDiagnosticsResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(CodeAnalysisDiagnosticsResources.DoNotUseCompilationGetSemanticModelMessage), CodeAnalysisDiagnosticsResources.ResourceManager, typeof(CodeAnalysisDiagnosticsResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(CodeAnalysisDiagnosticsResources.DoNotUseCompilationGetSemanticModelDescription), CodeAnalysisDiagnosticsResources.ResourceManager, typeof(CodeAnalysisDiagnosticsResources));
+
+        public static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            DiagnosticIds.DoNotUseCompilationGetSemanticModelRuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.MicrosoftCodeAnalysisCorrectness,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            description: s_localizableDescription,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            base.Initialize(context);
+        }
+
+        protected override DiagnosticAnalyzerSymbolAnalyzer? GetDiagnosticAnalyzerSymbolAnalyzer(CompilationStartAnalysisContext compilationContext, INamedTypeSymbol diagnosticAnalyzer, INamedTypeSymbol diagnosticAnalyzerAttribute)
```

Comment:
Your current implementation will flag invocation of this API even outside a DiagnosticAnalyzer type. Code in the symbol action callback of DiagnosticAnalyzerSymbolAnalyzer is the one that reduces the analysis to just inside diagnostic analyzer type: https://github.com/dotnet/roslyn-analyzers/blob/09f25eda649579b875952673cbdd3c8997b07ab0/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticAnalyzerCorrectnessAnalyzer.DiagnosticAnalyzerSymbolAnalyzer.cs#L27-L34

I think a correct approach here is to create a new type `DiagnosticAnalyzerSymbolStartAnalyzer` with methods `internal void AnalyzeSymbolStart(SymbolStartAnalysisContext symbolStartContext)` and `protected abstract void AnalyzeDiagnosticAnalyzer(SymbolStartAnalysisContext symbolStartContext);` analogous to the ones in `DiagnosticAnalyzerSymbolAnalyzer.cs`. You probably want to share the rest of the code between `DiagnosticAnalyzerSymbolAnalyzer` and `DiagnosticAnalyzerSymbolStartAnalyzer`. Then you can register the operation action on the symbol start context inside the override `AnalyzeDiagnosticAnalyzer(SymbolStartAnalysisContext symbolStartContext)`.

I will also be fine if you think it is better to get rid of `DiagnosticAnalyzerCorrectnessAnalyzer` type completely and extract some of its contents to just helper methods.